examples: Add single page (React) app with OAuth (#31534)
envoyproxy/envoy: docs/root/start/sandboxes/single-page-app.rst

    single-page-app-myhub-1       single-page-app-myhub       "/opt/myhub/app.py"                                        myhub     2 minutes ago   Up 2 minutes   0.0.0.0:7000->7000/tcp, :::7000->7000/tcp
    single-page-app-myhub-api-1   single-page-app-myhub-api   "/opt/myhub/app.py"                                        myhub-api 2 minutes ago   Up 2 minutes
    single-page-app-ui-1          single-page-app-ui          "/entrypoint.sh dev.sh"                                    ui        2 minutes ago   Up 2 minutes

.. _install_sandboxes_single_page_app_step_login:

Step 3: Browse to the dev app and login
***************************************

The development app should now be available on at http://localhost:10001 and provide a login button.

.. image:: /start/sandboxes/_static/spa-login.png
   :align: center

.. note::
   The dummy OAuth provider automatically trusts everyone as a hard-coded ``envoydemo`` user and redirects back to the app.

   In a real world scenario the provider would authenticate and authorize the user before proceeding.

> When a user clicks ``login`` it initiates the OAuth flow by calling the ``/authorize`` path:

Author: ```suggestion
When a user clicks ``login`` the app initiates the OAuth flow by calling the ``/login`` path:
```

current (ie these) docs are not correct it calls `/login` but it could be anything thats not explicitly excluded 

this requires some better explanation and perhaps defaults cc @mattklein123

   :emphasize-lines: 3-4
   :caption: :download:`envoy.yml <_include/single-page-app/envoy.yml>`

On successful authorization/authentication the user is redirected back to the app with the necessary OAuth code to proceed.

Envoy then uses this code with its client secret to generate an access token for the user which Envoy retains.

.. literalinclude:: _include/single-page-app/envoy.yml
   :language: yaml
   :lines: 50-60
   :linenos:
   :lineno-start: 50
   :emphasize-lines: 3-10
   :caption: :download:`envoy.yml <_include/single-page-app/envoy.yml>`

.. literalinclude:: _include/single-page-app/secrets/myhub-token-secret.yml
   :language: yaml
   :linenos:
   :emphasize-lines: 6
   :caption: :download:`myhub-token-secret.yml <_include/single-page-app/secrets/myhub-token-secret.yml>`

Once logged in, you should be able to make queries to the API using the OAuth credentials.

.. image:: /start/sandboxes/_static/spa-resources.png
   :align: center

.. warning::
   Envoy's OAuth implementation defaults to triggering the OAuth flow for all requests on the endpoint.

   This can readily trigger an OAuth flood as assets are requested, and doom loops when the OAuth flows fail.

>    This can be avoided by restricting the paths that are used by the OAuth flow.

Author: @mattklein123 flagging this - in initial dev i was constantly triggering a browser/server doom loop

i thought perhaps it was my implementation (to some extent this was true), but the more i debugged i realized its kinda endemic if the endpoint assumes all paths should trigger auth flow - relatedly there is some defensive code envoy-side to exit if it detects this (_mostly_ it didnt in my testing)

either way, default usage readily triggers this, and after quite a bit of testing/dev ive found it hard to prevent - not sure current implementation of react app/envoy config would be entirely reliable tbh - not sure how to prevent but i fwiw i do know how to trigger i think

Author: a bit more confident in the implementation here now - it inverts pass_through_matcher to only trigger oauth flow for the oauth endpoints - ill update the docs here - possibly we should add something to the filter docs

         }
       ]
     }
   ]

.. note::
   This setup configures Envoy to store the necessary files in memory.

   This may be a good fit for the single page app use case but would not scale well
   for many or large files.

.. tip::
   When you make changes to the javascript/typescript files rebuilding the app creates new routes to the
   compiled assets.

   In this case Envoy will update via xDS and use the newly routed assets.

   After making changes to assets that do not get a new route - eg ``index.html`` you
   should both rebuild the app and restart Envoy after.

   .. code-block:: console

      $ docker compose run --rm ui build.sh
      $ docker compose restart envoy

.. _install_sandboxes_single_page_app_step_production_browse:

Step 8: Browse to the production server
***************************************

You can browse to this server on https://localhost:10000

Unlike the development endpoint the production server is configured with TLS (self-signed) and Gzip compression, and
the assets are all served statically.

.. _install_sandboxes_single_page_app_step_github_oauth:

Step 9: Setup Github OAuth/API access
*************************************

.. tip::

>    Setup for `Github <github-link>`_ is explained here, but it should be easy to adapt these instructions for other providers.

Author: this link is confusing - similar with other gh links

The sandbox covers a number of Envoy's features, including:

- :ref:`direct_response <config_network_filters_direct_response>`
- :ref:`OAuth <config_http_filters_oauth>`
- Dynamic xDS filesystem updates
- Websocket proxy
- Gzip :ref:`compression <config_http_filters_compressor>`
- TLS/SNI up/downstream connection/termination
- Path/host rewrites

The app is built with `React <react_>`__ using `Vite <vite_>`__ and demonstrates OAuth authentication using
Envoy's :ref:`OAuth filter <config_http_filters_oauth>`.

This covers a scenario where we want OAuth to both authenticate the user and provide credentials
for further API interactions.

This is enabled by setting the OAuth configuration
:ref:`forward_bearer_token <envoy_v3_api_field_extensions.filters.http.oauth2.v3.OAuth2Config.forward_bearer_token>`
to ``true``

.. literalinclude:: _include/single-page-app/envoy.yml
   :language: yaml
   :lines: 36-40
   :linenos:
   :lineno-start: 36
   :emphasize-lines: 3
   :caption: :download:`envoy.yml <_include/single-page-app/envoy.yml>`

.. warning::
   Setting
   :ref:`forward_bearer_token <envoy_v3_api_field_extensions.filters.http.oauth2.v3.OAuth2Config.forward_bearer_token>`

>    means the provided access token will be forwarded to upstreams proxied by Envoy unless explicitly excluded.

Reviewer: This is a good warning, but what are the consequences? Presumably the point is you must trust upstream proxies, and if they issued the bearer token in the first place it's OK?

Author: if you have other proxies set up in envoy and you have the cookie - it will be sent to any upstreams that are not excluded

Reviewer: ```
... will be forwarded to all other upstreams for the same listener and filter chain, including those that did not issue the original token. 

This behavior can be avoided by configuring the :ref:`pass_through_matcher <envoy_v3_api_field_extensions.filters.http.oauth2.v3.OAuth2Config.pass_through_matcher>` or by disabling the OAuth2 filter with per-route configuration (link to https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#envoy-v3-api-field-extensions-filters-network-http-connection-manager-v3-httpfilter-disabled ).
```

Reviewer: I think the per-route configuration is more natural for Envoy, I might just drop the reference to pass_through_matcher TBH.

Reviewer: Although I wonder if this should be there to strip the headers from the client in. the case you don't want to pass through? In any case, having better per-route control over this filter would have made sense.

Reviewer: Actually, reading the code, pass_through_matcher isn't what you want. It will forward whatever the client sends, which includes bearer token. You'd want to remove the header somehow with RouteConfiguration mutations to avoid forwarding Authorization or bearer tokens to untrusted upstreams.

Author: `pass_through_matcher` needs to be set anyway - not sure re token - if i dont include `/hub*` path in the pass through it doesnt work

originally i had thought to add something about scrubbing headers, but then i noticed that it wasnt passing through - probs should check this again

Author: to clarify - if i dont add `hub*` to **inverted** `pass_through_matcher` - ie so it _is_ passed through it doesnt get the creds afaict

or put another way anything that is passed through gets no cookies/creds

Author: testing this further, it seems `pass_through_matcher` has no effect on creds they are passed through regardless

config doesnt need `hub.*` in the matcher

i will update/remove the ref to p_t_m here

i would rather not go too far into how to scrub the cookies/auth but would be good to ref something

This ignores all paths for OAuth other than:

- ``/authorize.*``
- ``/hub.*``
- ``/login``
- ``/logout``.

.. literalinclude:: _include/single-page-app/envoy.yml
   :language: yaml
   :lines: 37-46
   :linenos:
   :lineno-start: 37
   :emphasize-lines: 3-8
   :caption: :download:`envoy.yml <_include/single-page-app/envoy.yml>`

When a user clicks ``login`` the app initiates the OAuth flow by calling the ``/login`` path in Envoy.

This redirects the user to the OAuth provider for authorization/authentication with a further redirect link:

.. literalinclude:: _include/single-page-app/envoy.yml
   :language: yaml
   :lines: 34-39
   :linenos:
   :lineno-start: 34
   :emphasize-lines: 3-4
   :caption: :download:`envoy.yml <_include/single-page-app/envoy.yml>`

On successful authorization/authentication the user is redirected back via this link to the app with the necessary OAuth code to proceed:

.. literalinclude:: _include/single-page-app/envoy.yml
   :language: yaml
   :lines: 44-50
   :linenos:
   :lineno-start: 44
   :emphasize-lines: 3-5
   :caption: :download:`envoy.yml <_include/single-page-app/envoy.yml>`

> Envoy then uses this code with its client secret to generate an access token for the user, which Envoy then retains:

Reviewer: I'm not sure I understand this portion actually. The token is being retained in memory right? Is this on the response path? Kind of confused.

Author: this is how envoy/oauth works - after authorization user returns and makes a requiest including `code`

oauth server issues token to envoy after client makes request including `code`

wondering how to make this cleaerer - open to suggestions

Author: following up offline conv - the formal name is "authorization code" https://oauth.net/2/grant-types/authorization-code/

i will update terminology in pr

Author: done

.. literalinclude:: _include/single-page-app/envoy.yml
   :language: yaml
   :lines: 50-61
   :linenos:
   :lineno-start: 50
   :emphasize-lines: 3-10
   :caption: :download:`envoy.yml <_include/single-page-app/envoy.yml>`

.. literalinclude:: _include/single-page-app/secrets/myhub-token-secret.yml
   :language: yaml
   :linenos:
   :emphasize-lines: 6
   :caption: :download:`myhub-token-secret.yml <_include/single-page-app/secrets/myhub-token-secret.yml>`

Once logged in, you should be able to make queries to the API using the OAuth credentials:

.. image:: /start/sandboxes/_static/spa-resources.png
   :align: center

.. warning::
   Envoy's OAuth implementation defaults to triggering the OAuth flow for all paths on the endpoint.

   This can readily trigger an OAuth flood as assets are requested, and doom loops when the OAuth flows fail.

   This can be avoided by restricting the paths that are used by the OAuth flow.

   The sandbox example does this by inverting the
   :ref:`pass_through_matcher <envoy_v3_api_field_extensions.filters.http.oauth2.v3.OAuth2Config.pass_through_matcher>`

>    to only match on the required Oauth paths.

Author: ```suggestion
   to only match on the required OAuth paths.
```

         }
       },
       "response_headers_to_add": [
         {
           "header": {
             "key": "Content-Type",
             "value": "text/html"
           }
         }
       ]
     }
   ]

.. note::
   This setup configures Envoy to store the necessary files in memory.

   This may be a good fit for the single page app use case, but would not scale well
   for many or large files.

.. tip::
   When you make changes to the javascript/typescript files rebuilding the app creates new routes to the
   compiled assets.

   In this case Envoy will update via xDS and use the newly routed assets.

   If you make changes only to assets that do not get a new route - e.g. ``index.html`` - you
   should both rebuild the app and restart Envoy after:

   .. code-block:: console

      $ docker compose run --rm ui build.sh
      $ docker compose restart envoy

.. _install_sandboxes_single_page_app_step_production_browse:

Step 8: Browse to the production server
***************************************

You can browse to this server on https://localhost:10000

Unlike the development endpoint the production endpoint is configured with:

- TLS (self-signed)
- Gzip compression

> - Statically.served assets

Reviewer: ```suggestion
- Statically served assets
```

.. _install_sandboxes_single_page_app:

Single page React app (with OAuth)
==================================

.. sidebar:: Requirements

   .. include:: _include/docker-env-setup-link.rst

   :ref:`curl <start_sandboxes_setup_curl>`
        Used to make HTTP requests.

   :ref:`envsubst <start_sandboxes_setup_envsubst>`
        Used to interpolate environment vars in templates.

   :ref:`mkpasswd <start_sandboxes_setup_mkpasswd>`

>         Used to generate a ~random HMAC token.

Author: add `jq`

This ignores all paths for OAuth other than:

- ``/authorize.*``
- ``/hub.*``
- ``/login``
- ``/logout``.

.. literalinclude:: _include/single-page-app/envoy.yml
   :language: yaml
   :lines: 37-46
   :linenos:
   :lineno-start: 37
   :emphasize-lines: 3-8
   :caption: :download:`envoy.yml <_include/single-page-app/envoy.yml>`

When a user clicks ``login`` the app initiates the OAuth flow by calling the ``/login`` path in Envoy.

This redirects the user to the OAuth provider for authorization/authentication with a further redirect link:

.. literalinclude:: _include/single-page-app/envoy.yml
   :language: yaml
   :lines: 34-39
   :linenos:
   :lineno-start: 34
   :emphasize-lines: 3-4
   :caption: :download:`envoy.yml <_include/single-page-app/envoy.yml>`

On successful authorization/authentication the user is redirected back via this link to the app with the necessary OAuth
`authorization code <oauth-auth-code_>`__ to proceed:

.. literalinclude:: _include/single-page-app/envoy.yml
   :language: yaml
   :lines: 44-50
   :linenos:
   :lineno-start: 44
   :emphasize-lines: 3-5
   :caption: :download:`envoy.yml <_include/single-page-app/envoy.yml>`

> Envoy then uses this authorization code with its client secret to generate an access token for the user, which Envoy then retains:

Reviewer: I don't think it retains it. I think it turns it into a client cookie, see https://github.com/envoyproxy/envoy/blob/12210e53db6cea79b58ee027f047faa356e1e392/source/extensions/filters/http/oauth2/filter.cc#L565.

Author: im pretty sure envoy keeps something - if i delete my cooking and refresh the page it reappears - again i should probs double check this - i was basing this of what i observed not checking source

Author: k - checking this more throroughly - it doesnt retain anything afaict

i tracked down why this seemed to be the case - i was using http://localhost:10001/hub/user directly to check credentials - but this endpoint triggers the oauth flow so i was simply reauthing

ill update (still not clear what the hmac used for but its not for storing anything in envoy afaict)

      :caption: :download:`envoy.yml <_include/single-page-app/envoy.yml>`

.. _install_sandboxes_single_page_app_step_api:

Step 4: Make API queries
************************

For the sandbox app,
:ref:`forward_bearer_token <envoy_v3_api_field_extensions.filters.http.oauth2.v3.OAuth2Config.forward_bearer_token>`
is set, and so Envoy also passes the acquired access token back to the user as a cookie:

.. image:: /start/sandboxes/_static/spa-cookies.png
   :align: center

This cookie is then passed through Envoy in any subsequent requests to the proxied Myhub API:

.. literalinclude:: _include/single-page-app/envoy.yml
   :language: yaml
   :lines: 74-86
   :linenos:
   :lineno-start: 74
   :emphasize-lines: 3-11
   :caption: :download:`envoy.yml <_include/single-page-app/envoy.yml>`

.. _install_sandboxes_single_page_app_step_reload:

Step 5: Live reload code changes
********************************

With your browser open on http://localhost:10001 make some change to the UI.

For example, you might change the page title:

.. code-block:: console

   $ sed -i s/Envoy\ single\ page\ app\ example/DEV\ APP/g ui/index.html

The page should automatically refresh.

> Likewise any changes to the components in ``ui/src/...`` should automatically reload in

Author: ```suggestion
Likewise any changes to the app components in ``ui/src/...`` should automatically reload in
```

         }
       },
       "response_headers_to_add": [
         {
           "header": {
             "key": "Content-Type",
             "value": "text/javascript"
           }
         }
       ]
     },
     {
       "match": {
         "path": "/myhub.svg"
       },
       "direct_response": {
         "status": 200,
         "body": {
           "filename": "/var/www/html/myhub.svg"
         }
       },
       "response_headers_to_add": [
         {
           "header": {
             "key": "Content-Type",
             "value": "image/svg+xml"
           }
         }
       ]
     },
     {
       "match": {
         "prefix": "/"
       },
       "direct_response": {
         "status": 200,
         "body": {
           "filename": "/var/www/html/index.html"
         }
       },
       "response_headers_to_add": [
         {
           "header": {
             "key": "Content-Type",
             "value": "text/html"
           }
         }
       ]
     }
   ]

.. note::
   This setup configures Envoy to store the necessary files in memory.

   This may be a good fit for the single page app use case, but would not scale well
   for many or large files.

.. tip::

>    When you make changes to the javascript/typescript files rebuilding the app creates new routes to the

Author: ```suggestion
   When you make changes to the Typescript files rebuilding the app creates new routes to the
```

   compiled assets.

   In this case Envoy will update via xDS and use the newly routed assets.

   If you make changes only to assets that do not get a new route - e.g. ``index.html`` - you
   should both rebuild the app and restart Envoy after:

   .. code-block:: console

      $ docker compose run --rm ui build.sh
      $ docker compose restart envoy

.. _install_sandboxes_single_page_app_step_production_browse:

Step 8: Browse to the production server
***************************************

You can browse to this server on https://localhost:10000

Unlike the development endpoint the production endpoint is configured with:

- TLS (self-signed)
- Gzip compression
- Statically.served assets

.. _install_sandboxes_single_page_app_step_github_oauth:

Step 9: Setup Github OAuth/API access
*************************************

.. tip::
   Setup for `Github <github_>`__ is explained in this sandbox, but it should be easy to adapt these instructions for other providers.

> You will need to set up either an `Github OAuth or full app <github-oauth_>`__. The latter provides

Author: ```suggestion
You will need to set up either a `Github OAuth or full app <github-oauth_>`__. The latter provides
```

more control and is generally preferable.

> This can be done either at the `user <https://github.com/settings/developers>`_ or organization levels:

Author: add ref link

The sandbox covers a number of Envoy's features, including:

- :ref:`direct_response <config_network_filters_direct_response>`
- :ref:`OAuth <config_http_filters_oauth>`
- Dynamic xDS filesystem updates
- Websocket proxy
- Gzip :ref:`compression <config_http_filters_compressor>`
- TLS/SNI up/downstream connection/termination
- Path/host rewrites

The app is built with `React <react_>`__ using `Vite <vite_>`__ and demonstrates OAuth authentication using
Envoy's :ref:`OAuth filter <config_http_filters_oauth>`.

This covers a scenario where we want OAuth to both authenticate the user and provide credentials
for further API interactions.

This is enabled by setting the OAuth configuration
:ref:`forward_bearer_token <envoy_v3_api_field_extensions.filters.http.oauth2.v3.OAuth2Config.forward_bearer_token>`
to ``true``

.. literalinclude:: _include/single-page-app/envoy.yml
   :language: yaml
   :lines: 36-40
   :linenos:
   :lineno-start: 36
   :emphasize-lines: 3
   :caption: :download:`envoy.yml <_include/single-page-app/envoy.yml>`

.. warning::
   Setting
   :ref:`forward_bearer_token <envoy_v3_api_field_extensions.filters.http.oauth2.v3.OAuth2Config.forward_bearer_token>`

>    means the provided access token will be forwarded to upstreams proxied by Envoy unless explicitly excluded.

Author: ill work this in

>    means the provided access token will be forwarded to upstreams proxied by Envoy unless explicitly excluded.

Author: i may be wrong but i think the important thing is removing the cookie/header - not sure that enabling/disabling the oauth filter would achieve this or help - besides in this situation you would need to disable it for given upstreams anyway (whether by pass_through or disabling the filter) just to get it to work correctly

ive updated on this basis

   This can be avoided by disabling the OAuth2 filter with
   :ref:`per-route configuration <envoy_v3_api_field_extensions.filters.network.http_connection_manager.v3.HttpFilter.disabled>`.

A dummy "Myhub" backend is provided with a minimal OAuth provider and API for use in the example.

> Setup is provided to :ref:`build and update the app for production use <install_sandboxes_single_page_app_step_production_build>`,

Author: its hard not to call it `production` becase that is what js/node calls it when you build the assets rather than serve them

eg - it reads `.env.production` in that circumstance

`authorization code <oauth-auth-code_>`__ to proceed:

.. literalinclude:: _include/single-page-app/envoy.yml
   :language: yaml
   :lines: 44-50
   :linenos:
   :lineno-start: 44
   :emphasize-lines: 3-5
   :caption: :download:`envoy.yml <_include/single-page-app/envoy.yml>`

Envoy then uses this authorization code with its client secret to confirm authorization and obtain an access token for the user:

.. literalinclude:: _include/single-page-app/envoy.yml
   :language: yaml
   :lines: 50-61
   :linenos:
   :lineno-start: 50
   :emphasize-lines: 3-10
   :caption: :download:`envoy.yml <_include/single-page-app/envoy.yml>`

.. literalinclude:: _include/single-page-app/secrets/myhub-token-secret.yml
   :language: yaml
   :linenos:
   :emphasize-lines: 6
   :caption: :download:`myhub-token-secret.yml <_include/single-page-app/secrets/myhub-token-secret.yml>`

Once logged in, you should be able to make queries to the API using the OAuth credentials:

.. image:: /start/sandboxes/_static/spa-resources.png
   :align: center

.. warning::
   Envoy's OAuth implementation defaults to triggering the OAuth flow for all paths on the endpoint.

>    This can readily trigger an OAuth flood as assets are requested, and doom loops when the OAuth flows fail.

Reviewer: Not sure I fully follow - shouldn't the access token creation obtain sufficient scope to be able to reuse the OAuth cookie in future accesses to other paths?

Author: if you dont restrict what triggers the oauth flow then its triggered on the assets

this can cause redirect loops which take over the client (browser) machine and flood the proxy

not knowing this when i started i really struggled to get anything working so i think we need to say something - here im trying to highlight the strategy/config used to avoid that (ie inverse matching)

   The sandbox example does this by inverting the
   :ref:`pass_through_matcher <envoy_v3_api_field_extensions.filters.http.oauth2.v3.OAuth2Config.pass_through_matcher>`
   to only match on the required OAuth paths.

.. tip::
   The Myhub OAuth provider does not provide an expiry for issued credentials. Likewise Github may or may
   not depending on configuration. This is valid in terms of the OAuth2 specification.

   If the authorization provider does not include an expiry, Envoy will, by default, fail the authentication.

   This can be resolved by setting
   :ref:`default_expires_in <envoy_v3_api_field_extensions.filters.http.oauth2.v3.OAuth2Config.default_expires_in>`:

   .. literalinclude:: _include/single-page-app/envoy.yml
      :language: yaml
      :lines: 33-37
      :linenos:
      :lineno-start: 33
      :emphasize-lines: 3
      :caption: :download:`envoy.yml <_include/single-page-app/envoy.yml>`

.. _install_sandboxes_single_page_app_step_api:

Step 5: Make API queries
************************

For the sandbox app,
:ref:`forward_bearer_token <envoy_v3_api_field_extensions.filters.http.oauth2.v3.OAuth2Config.forward_bearer_token>`

> is set, and so Envoy also passes the acquired access token back to the user as a cookie:

Reviewer: It's my understanding that `forward_bearer_token` governs Envoy to upstream, rather than Envoy to client cookie handling.

Reviewer: The cookie is written in any case AFAICT.

Author: let me check that again - iirc the browser doesnt get it in that case but i can check again (also wondering what point of cookie would be if its not forwarded)

Reviewer: Actually, even more interesting is to follow https://github.com/envoyproxy/envoy/blob/c275fa8d4dfa44fa377b99b4b4eb931f8011c40e/source/extensions/filters/http/oauth2/filter.cc#L636. It sets the OAuth cookie unconditionally, but then an additional bearer cookies for the forward bearer token option. The unconditinoal OAuth cookie though is missing the access token without this option. TIL.

I think what you write is roughly correct then.

Reviewer: So I think the cookie exists for the proxy authorization only when not setting `forward_bearer_token`. The cookie is extended with the access token for future accesses when forwarding the token, as Envoy is not remembering it. So, what you have it correct, even though there are cookies in both cases.
